### PR TITLE
VDS-175 Move customer segments to separate module

### PR DESCRIPTION
**Pull request by trueboroda**

### CODE REVIEW

**Comment by mvktsk** [comment]
Review task https://virtocommerce.atlassian.net/browse/VP-6301 has been created

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Web/module.manifest [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    <version-tag></version-tag>
+    <platformVersion>3.0.0</platformVersion>
+    <dependencies>
+        <dependency id="VirtoCommerce.Core" version="3.0.0" /> 
+    </dependencies>
+
+    <title>VirtoCommerce.DemoCustomerSegmentsModule</title>
+    <description>This module adds the customer segments functionality to VirtoCommerce platform.</description>
+    <authors>
+        <author>asvishnyakov</author>
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Web/module.manifest:14]: Please use full developer names

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Web/module.manifest [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<module>
+    <id>VirtoCommerce.DemoCustomerSegmentsModule</id>
+    <version>1.0.0</version>
+    <version-tag></version-tag>
+    <platformVersion>3.0.0</platformVersion>
+    <dependencies>
+        <dependency id="VirtoCommerce.Core" version="3.0.0" />
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Web/module.manifest:8]: Where is marketing module dependency?

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Scripts/resources/customerSegments.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+angular.module('virtoCommerce.DemoCustomerSegmentsModule')
+.factory('virtoCommerce.DemoCustomerSegmentsModule.customerSegmentsApi', ['$resource', function ($resource) {
+    return $resource('api/demo/customersegments/:id', {}, {
+        new: { method: 'GET', url: 'api/demo/customersegments/new' },
+        update: { method: 'PUT' },
+        delete: { method: 'DELETE' },
+        search: { method: 'POST', url: 'api/demo/customersegments/search' },
+        preview: { method: 'POST', url: 'api/demo/customersegments/preview', isArray: true }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Scripts/resources/customerSegments.js:8]: We don't have this endpoint anymore, isn't it?

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Scripts/blades/customerSegments-preview.tpl.html [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+<script type="text/ng-template" id="customerSegment-preview.row.html">
+    <div ng-repeat="(colRenderIndex, col) in colContainer.renderedColumns track by col.uid" ui-grid-one-bind-id-grid="rowRenderIndex + '-' + col.uid + '-cell'" class="ui-grid-cell" ng-class="{'ui-grid-row-header-cell': col.isRowHeader, '__selected': row.entity.id===grid.appScope.selectedNodeId }" role="{{col.isRowHeader ? 'rowheader' : 'gridcell' }}" ui-grid-cell></div>
+</script>
+<script type="text/ng-template" id="customerSegment-preview-icon.cell.html">
+    <div class="ui-grid-cell-contents">
+        <div class="product-img" ng-attr-title="{{COL_FIELD}}">
+            <i class="table-ico fa" ng-class="row.entity._memberTypeIcon ? row.entity._memberTypeIcon : 'fa-user'"></i>
+        </div>
+    </div>
+</script>
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Scripts/blades/customerSegments-preview.tpl.html:44]: Add empty string in the end of file

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Module.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            permissionsProvider.RegisterPermissions(ModuleConstants.Security.Permissions.AllPermissions.Select(x =>
+                new Permission()
+                {
+                    GroupName = "VirtoCommerceDemoCustomerSegmentsModule",
+                    ModuleId = ModuleInfo.Id,
+                    Name = x
+                }).ToArray());
+
+            // Ensure that any pending migrations are applied
+            using (var serviceScope = appBuilder.ApplicationServices.CreateScope())
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Module.cs:59]: Union using declarations

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Controllers/Api/DemoCustomerSegmentController.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        /// <summary>
+        /// Search customer segments by specified search criteria.
+        /// </summary>
+        /// <param name="criteria">Search criteria</param>
+        /// <returns>Search result with total number of found customer segments and all found customer segments.</returns>
+        [HttpPost]
+        [Route("search")]
+        [Authorize(ModuleConstants.Security.Permissions.Read)]
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Controllers/Api/DemoCustomerSegmentController.cs:101]: Should we use here marketing permissions?

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Core/VirtoCommerce.DemoCustomerSegmentsModule.Core.csproj [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="VirtoCommerce.CoreModule.Core" Version="3.8.0" />
```
Comment (asvishnyakov) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Core/VirtoCommerce.DemoCustomerSegmentsModule.Core.csproj:8]: Should be 3.7.0

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Core/VirtoCommerce.DemoCustomerSegmentsModule.Core.csproj [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    <PackageReference Include="VirtoCommerce.Platform.Core" Version="3.16.0" />
```
Comment (asvishnyakov) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Core/VirtoCommerce.DemoCustomerSegmentsModule.Core.csproj:9]: Should be 3.29.0

**Review by asvishnyakov — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Data/VirtoCommerce.DemoCustomerSegmentsModule.Data.csproj [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    <Folder Include="Handlers\" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="3.1.2">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="VirtoCommerce.CoreModule.Core" Version="3.8.0" />
+    <PackageReference Include="VirtoCommerce.Platform.Data" Version="3.0.0" />
```
Comment (asvishnyakov) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Data/VirtoCommerce.DemoCustomerSegmentsModule.Data.csproj:22]: Update versions to used on environment too

**Review by asvishnyakov — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Core/VirtoCommerce.DemoCustomerSegmentsModule.Core.csproj (hunk 7 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoCustomerSegmentsModule.Core/VirtoCommerce.DemoCustomerSegmentsModule.Core.csproj]: Fixed

**Review by asvishnyakov — APPROVED** [review]

**Review by pushnitsa — DISMISSED** [review]

**Review by ilyawzrd — APPROVED** [review]

**Review by ilyawzrd — DISMISSED** [review]

**Review by asvishnyakov — APPROVED** [review]

**Review by ilyawzrd — APPROVED** [review]

**Pull request merged**